crates/ruff_linter/src/rules/ruff/rules/sort_dunder_slots.rs

```diff
@@ -0,0 +1,272 @@
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+enum SpecialClassDunder {
+    Slots,
+    MatchArgs,
+}
+
+impl Display for SpecialClassDunder {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let string = match self {
+            Self::MatchArgs => "__match_args__",
+            Self::Slots => "__slots__",
+        };
+        write!(f, "{string}")
+    }
+}
+
+/// Sort a `__slots__`/`__match_args__` definition
+/// represented by a `StmtAssign` AST node.
+/// For example: `__slots__ = ["b", "c", "a"]`.
+pub(crate) fn sort_dunder_slots_assign(
+    checker: &mut Checker,
+    ast::StmtAssign { value, targets, .. }: &ast::StmtAssign,
+) {
+    if let [expr] = targets.as_slice() {
+        sort_dunder_slots(checker, expr, value);
+    }
+}
+
+/// Sort a `__slots__`/`__match_args__` definition
+/// represented by a `StmtAnnAssign` AST node.
+/// For example: `__slots__: list[str] = ["b", "c", "a"]`.
+pub(crate) fn sort_dunder_slots_ann_assign(checker: &mut Checker, node: &ast::StmtAnnAssign) {
+    if let Some(value) = &node.value {
+        sort_dunder_slots(checker, &node.target, value);
+    }
+}
+
+/// Sort a tuple, list, dict or set that defines `__slots__`
+/// or `__match_args__` in a class scope.
+///
+/// This routine checks whether the display is sorted, and emits a
+/// violation if it is not sorted. If the tuple/list/set was not sorted,
+/// it attempts to set a `Fix` on the violation.
+fn sort_dunder_slots(checker: &mut Checker, target: &ast::Expr, node: &ast::Expr) {
+    let ast::Expr::Name(ast::ExprName { id, .. }) = target else {
+        return;
+    };
+
+    let dunder_kind = match id.as_str() {
+        "__slots__" => SpecialClassDunder::Slots,
+        "__match_args__" => SpecialClassDunder::MatchArgs,
+        _ => return,
+    };
+
+    // We're only interested in `__slots__`/`__match_args__` in the class scope
+    let ScopeKind::Class(ast::StmtClassDef {
+        name: class_name, ..
+    }) = checker.semantic().current_scope().kind
+    else {
+        return;
+    };
+
+    let Some(elts_analysis) = NodeAnalysis::new(node, dunder_kind) else {
+        return;
+    };
+
+    let sort_classification = SortClassification::of_elements(&elts_analysis.elts, natord::compare);
+    if sort_classification.is_not_a_list_of_string_literals() || sort_classification.is_sorted() {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(
+        UnsortedDunderSlots {
+            class_name: class_name.to_string(),
+            class_variable: dunder_kind,
+        },
+        elts_analysis.range,
+    );
+
+    if let SortClassification::UnsortedAndMaybeFixable { items } = sort_classification {
+        if let Some(fix) = create_fix(&elts_analysis, &items, checker) {
+            diagnostic.set_fix(fix);
+        }
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+#[derive(Debug)]
+struct NodeAnalysis<'a> {
```

Comment:
Nit: The name here isn't telling me much. What kind of analysis is it? Is it like the `SortDunderAnalysis`? Or can we give that structure another name, because the structure isn't the analysis. What does it represent?

Reply:
Pushed a rename and some more docs in cf6ca7564eb268ab3098c3a11d3787c36501e40f -- any better?

I really struggled with naming this the other day, but coming back to it after a few days made it seem obvious :)